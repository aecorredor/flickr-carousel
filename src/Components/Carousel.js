import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { css, cx } from 'emotion';
import isEmpty from 'lodash.isempty';

const flexColumn = css`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const arrowContainer = css`
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
`;

const photoItem = css`
  min-width: 15rem;
  margin: 0.5rem;
  background-position: 50% 50%;
  background-repeat: no-repeat;
  background-size: contain;
  outline: none;
`;

const arrowStyle = css`
  cursor: pointer;
  outline: none;
`;
const arrowClasses = cx('large material-icons', arrowStyle);

export default class Carousel extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedPhoto: this.props.selectedPhoto,
    };

    this.handleLeftArrowClick = this.handleLeftArrowClick.bind(this);
    this.handleRightArrowClick = this.handleRightArrowClick.bind(this);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.selectedPhoto !== this.props.selectedPhoto) {
      this.setState({ selectedPhoto: this.props.selectedPhoto });
    }
  }

  handleLeftArrowClick() {
    if (this.state.selectedPhoto === 0) {
      const photos = Object.values(this.props.photos);
      return this.setState({ selectedPhoto: photos.length - 1 }, this.handlePhotoChange);
    }

    this.setState(
      prevState => ({ selectedPhoto: prevState.selectedPhoto - 1 }),
      this.handlePhotoChange,
    );
  }

  handleRightArrowClick() {
    const photos = Object.values(this.props.photos);

    if (this.state.selectedPhoto === photos.length - 1) {
      return this.setState({ selectedPhoto: 0 }, this.handlePhotoChange);
    }

    this.setState(
      prevState => ({ selectedPhoto: prevState.selectedPhoto + 1 }),
      this.handlePhotoChange,
    );
  }

  handlePhotoChange() {
    if (this.props.handlePhotoChange) {
      this.props.handlePhotoChange(this.state.selectedPhoto);
    }
  }

  renderNavArrows() {
    return (
      <div className={arrowContainer}>
        <i
          role="button"
          tabIndex={0}
          onClick={this.handleLeftArrowClick}
          className={arrowClasses}
        >
          keyboard_arrow_left
        </i>
        <i
          role="button"
          tabIndex={0}
          onClick={this.handleRightArrowClick}
          className={arrowClasses}
        >
          keyboard_arrow_right
        </i>
      </div>
    );
  }

  render() {
    if (isEmpty(this.props.photos)) {
      return null;
    }

    const selectedPhoto = this.props.photos[this.state.selectedPhoto] || {};
    const mainPhoto = css`
      ${flexColumn};
      ${photoItem};
      width: 90%;
      height: 350px;
      margin: 0;
      background-image: url('${selectedPhoto.src}');
    `;

    return (
      <div className={mainPhoto}>
        {this.renderNavArrows()}
      </div>
    );
  }
}

Carousel.propTypes = {
  selectedPhoto: PropTypes.number,
  photos: PropTypes.objectOf(
    PropTypes.shape({
      src: PropTypes.string,
    }),
  ),
  handlePhotoChange: PropTypes.func,
};

Carousel.defaultProps = {
  selectedPhoto: 0,
  photos: { 0: { src: '' } },
  handlePhotoChange: null,
};
