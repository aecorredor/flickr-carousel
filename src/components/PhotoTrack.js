import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { css } from 'emotion';
import isEmpty from 'lodash.isempty';

const trackContainer = css`
  width: 64rem;
  overflow: hidden;
  background-color: #8e44ad;
`;

const innerTrack = css`
  display: flex;
  min-height: 200px;
  min-width: 100vw;
  transition: transform 0.5s ease-in;
`;

const photoItem = css`
  min-width: 15rem;
  margin: 0.5rem;
  background-position: 50% 50%;
  background-repeat: no-repeat;
  background-size: cover;
  outline: none;
`;

export default class PhotoTrack extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedPhoto: this.props.selectedPhoto,
    };

    this.handlePhotoClick = this.handlePhotoClick.bind(this);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.selectedPhoto !== this.props.selectedPhoto) {
      this.setState({ selectedPhoto: this.props.selectedPhoto });
    }
  }

  handlePhotoClick(e) {
    this.setState({ selectedPhoto: Number(e.target.dataset.index) }, this.handlePhotoChange);
  }

  handlePhotoChange() {
    if (this.props.handlePhotoChange) {
      this.props.handlePhotoChange(this.state.selectedPhoto);
    }
  }

  renderPhotoItem(photo) {
    const photoClass = css`
      ${photoItem};
      background-image: url('${photo.src}');
      border: ${this.state.selectedPhoto === photo.index ? '2px solid orange' : 'none'};
    `;

    return (
      <button
        key={photo.index}
        type="button"
        data-index={photo.index}
        onClick={this.handlePhotoClick}
        className={photoClass}
      />
    );
  }

  render() {
    if (isEmpty(this.props.photos)) {
      return null;
    }

    const selectedPhoto = this.props.photos[this.state.selectedPhoto] || {};
    const innerTrackClasses = css`
      ${innerTrack};
      transform: translateX(${selectedPhoto.trackPosition}rem);
    `;
    const photos = Object.values(this.props.photos);

    return (
      <div className={trackContainer}>
        <div className={innerTrackClasses}>
          {photos.map(photo => this.renderPhotoItem(photo))}
        </div>
      </div>
    );
  }
}

PhotoTrack.propTypes = {
  selectedPhoto: PropTypes.number,
  photos: PropTypes.objectOf(
    PropTypes.shape({
      src: PropTypes.string,
    }),
  ),
  handlePhotoChange: PropTypes.func,
};

PhotoTrack.defaultProps = {
  selectedPhoto: 0,
  photos: { 0: { src: '' } },
  handlePhotoChange: null,
};
