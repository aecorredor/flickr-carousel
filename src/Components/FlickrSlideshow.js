import React, { Component } from 'react';
import { css } from 'emotion';
import debounce from 'lodash.debounce';
import get from 'lodash.get';

import FlickrServices from '../services/flickr-services';

import Spinner from './Spinner';
import Carousel from './Carousel';
import PhotoTrack from './PhotoTrack';

const flexColumn = css`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const container = css`
  ${flexColumn};
  width: 80%;
  margin: 0 auto;
`;

const spinnerContainer = css`
  ${flexColumn};
  height: 566px;
`;

const searchField = css`
  width: 20rem;
`;

export default class FlickrSlideshow extends Component {
  constructor(props) {
    super(props);

    this.state = {
      searchingPhotos: false,
      searchTerm: '',
      photos: {},
      selectedPhoto: null,
      keyPressed: null,
    };

    this.handleSearchChange = this.handleSearchChange.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
    this.handlePhotoChange = this.handlePhotoChange.bind(this);
    this.debouncedSearchPhotos = debounce(this.searchPhotos, 750, { trailing: true });
  }

  handleSearchChange(e) {
    const searchTerm = e.target.value;
    const parsedSearchTerm = searchTerm.trim();
    const pressedKeyIsBackspace = this.state.keyPressed === 'Backspace';


    this.setState({ searchTerm }, () => {
      if (!parsedSearchTerm || parsedSearchTerm.length < 2 || pressedKeyIsBackspace) {
        return;
      }

      this.debouncedSearchPhotos(parsedSearchTerm);
    });
  }

  handleKeyPress(e) {
    this.setState({ keyPressed: e.key });
  }

  handlePhotoChange(newSelectedPhoto) {
    this.setState({ selectedPhoto: newSelectedPhoto });
  }

  async searchPhotos(searchTerm) {
    this.setState({ searchingPhotos: true });
    let results = [];

    try {
      const { body = {} } = await FlickrServices.searchPhotos(searchTerm);
      const rawPhotos = get(body, 'photos.photo', []);
      const photoSizesReq = rawPhotos.map(rawPhoto => FlickrServices.getSizes(rawPhoto.id));
      const photoSizes = await Promise.all(photoSizesReq);

      results = photoSizes.map((photoSize) => {
        const { body: photoBody = {} } = photoSize;
        const { sizes: { size = [] } } = photoBody;
        // From seeing the flickr API results, the last element of the size array is always
        // the original photo size.
        const selectedSize = size[size.length - 1];

        return selectedSize.source;
      });
    } catch (error) {
      console.log(`error fetching photos: ${error}`);
    } finally {
      this.buildCarouselData(results);
    }
  }

  buildCarouselData(results) {
    let trackPosition = 24; // Positions first picture in the middle of the track.

    const photos = results.reduce((acc, src, index) => {
      acc[index] = {
        src,
        trackPosition,
        index,
      };
      trackPosition -= 16;

      return acc;
    }, {});

    this.setState({ photos, searchingPhotos: false, selectedPhoto: 0 });
  }

  renderSpinner() {
    return (
      <div className={spinnerContainer}>
        <Spinner />
      </div>
    );
  }

  renderSearchInput() {
    return (
      <div className={searchField}>
        <input
          placeholder="Search photos..."
          value={this.state.searchTerm}
          onChange={this.handleSearchChange}
          onKeyDown={this.handleKeyPress}
        />
      </div>
    );
  }

  renderSlideshow() {
    return (
      <div className={flexColumn}>
        <Carousel
          photos={this.state.photos}
          selectedPhoto={this.state.selectedPhoto}
          handlePhotoChange={this.handlePhotoChange}
        />
        <PhotoTrack
          photos={this.state.photos}
          selectedPhoto={this.state.selectedPhoto}
          handlePhotoChange={this.handlePhotoChange}
        />
      </div>
    );
  }

  render() {
    return (
      <div className={container}>
        {this.renderSearchInput()}

        {
          this.state.searchingPhotos ?
            this.renderSpinner() :
            this.renderSlideshow()
        }
      </div>
    );
  }
}
