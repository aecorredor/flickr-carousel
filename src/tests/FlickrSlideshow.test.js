import React from 'react';
import FlickrSlideshow from '../components/FlickrSlideshow';
import Spinner from '../components/Spinner';
import Carousel from '../components/Carousel';
import PhotoTrack from '../components/PhotoTrack';

const searchTerm = 'nintendo';
const photos = {
  0: {
    src: 'example.jpg',
  },
  1: {
    src: 'example1.jpg',
  },
};

describe('<FlickrSlideShow />', () => {
  let wrapper;
  const onChangeSpy = jest.spyOn(FlickrSlideshow.prototype, 'handleSearchChange');
  const onKeyPressSpy = jest.spyOn(FlickrSlideshow.prototype, 'handleKeyPress');

  beforeEach(() => {
    wrapper = shallow(<FlickrSlideshow />);
  });

  it('should render search input correctly', () => {
    const searchInput = wrapper.childAt(0);

    expect(searchInput).toMatchSnapshot();
  });

  it('should handle search input onChange event', () => {
    const searchInput = wrapper.find('input').at(0);

    searchTerm.split('').reduce((acc, letter) => {
      const newValue = `${acc}${letter}`;

      searchInput.simulate('change', {
        target: { value: newValue },
      });

      return newValue;
    }, '');

    expect(onChangeSpy).toHaveBeenCalledTimes(searchTerm.length);
    expect(wrapper.state().searchTerm).toBe(searchTerm);
  });

  it('should handle search input keyPress event', () => {
    const searchInput = wrapper.find('input').at(0);

    searchTerm.split('').forEach((letter) => {
      searchInput.simulate('keydown', { key: letter });

      expect(onKeyPressSpy).toHaveBeenCalled();
      expect(wrapper.state().keyPressed).toBe(letter);
    });
  });

  it('should only render a spinner while searching', () => {
    wrapper.setState({ searchingPhotos: true }, () => {
      expect(wrapper.find(Spinner)).toHaveLength(1);
      expect(wrapper.find(Carousel)).toHaveLength(0);
      expect(wrapper.find(PhotoTrack)).toHaveLength(0);
    });
  });

  it('should initially only render the search input', () => {
    expect(wrapper.find('input')).toHaveLength(1);
    expect(wrapper.find(Spinner)).toHaveLength(0);
    expect(wrapper.find(Carousel)).toHaveLength(0);
    expect(wrapper.find(PhotoTrack)).toHaveLength(0);
  });

  it('should render a carousel when photos exist', () => {
    wrapper.setState({ photos, selectedPhoto: 0 }, () => {
      const carousel = wrapper.find(Carousel).first();

      expect(carousel.props().photos).toBe(photos);
      expect(carousel.props().selectedPhoto).toBe(0);
    });
  });

  it('should render a photo track when photos exist', () => {
    wrapper.setState({ photos, selectedPhoto: 0 }, () => {
      const photoTrack = wrapper.find(PhotoTrack).first();

      expect(photoTrack.props().photos).toBe(photos);
      expect(photoTrack.props().selectedPhoto).toBe(0);
    });
  });
});
