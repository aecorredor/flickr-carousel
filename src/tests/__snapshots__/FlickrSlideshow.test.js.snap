// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlickrSlideShow /> should render search input correctly 1`] = `
.emotion-0 {
  width: 20rem;
}

<div
  className="emotion-0"
>
  <input
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Search photos..."
    value=""
  />
</div>
`;
